### PR TITLE
Refactor InitialTopologyState logic to exist at bootstrapper level instead of in peer bootstrapper source

### PR DESCRIPTION
**Pull request by richardartoul**
- [X] Move InitialTopologyState logic to shared bootstrap code and then refactor peer bootstrapper to use new shared logic
- [X] Adjust all interfaces to accomodate
- [X] Fix all unit / integration tests

### CODE REVIEW

File: src/dbnode/storage/bootstrap/process.go [hunk 1]
```diff
@@ -81,21 +88,69 @@ func (b *bootstrapProcessProvider) Provide() (Process, error) {
...
+		log:                  b.log,
+		bootstrapper:         bootstrapper,
+		initialTopologyState: initialTopologyState,
 	}, nil
 }
 
+func (b *bootstrapProcessProvider) newInitialTopologyState() (*TopologyState, error) {
+	session, err := b.processOpts.AdminClient().DefaultAdminSession()
+	if err != nil {
+		return &TopologyState{}, err
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/process.go]: nit: maybe just `return nil, err`?

File: src/dbnode/storage/bootstrap/process.go [hunk 2]
```diff
@@ -81,21 +88,69 @@ func (b *bootstrapProcessProvider) Provide() (Process, error) {
...
+	}
+
+	topoMap, err := session.TopologyMap()
+	if err != nil {
+		return &TopologyState{}, err
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/process.go]: here too

File: src/dbnode/storage/bootstrap/process.go [hunk 3]
```diff
@@ -302,5 +357,30 @@ func (b bootstrapProcess) newRunOptions() RunOptions {
 	return NewRunOptions().
 		SetCacheSeriesMetadata(
 			b.processOpts.CacheSeriesMetadata(),
-		)
+		).
+		SetInitialTopologyState(b.initialTopologyState)
+}
+
+// TopologyState represents a snapshot of the state of the topology at a
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/process.go]: nit: mind moving this to `topology.State`

File: src/dbnode/storage/bootstrap/process.go [hunk 4]
```diff
@@ -302,5 +357,30 @@ func (b bootstrapProcess) newRunOptions() RunOptions {
...
+// given moment.
+type TopologyState struct {
+	Self             string
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/process.go]: could you use a `topology.Host` for self too? and would Origin be more consistent a name?

File: src/dbnode/storage/bootstrap/types.go [hunk 5]
```diff
@@ -94,6 +104,14 @@ type RunOptions interface {
...
 	// provider should cache series metadata between runs.
 	CacheSeriesMetadata() bool
+
+	// SetInitialTopologyState sets the initial topology state as it was
+	// measured before the bootstrap process began.
+	SetInitialTopologyState(value *TopologyState) RunOptions
+
+	// InitialTopologyState returns the initial topology as it was measured
+	// before the bootstrap process began.
+	InitialTopologyState() *TopologyState
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/types.go]: any reason to have this be a pointer type?

File: src/dbnode/integration/commitlog_bootstrap_index_test.go [hunk 6]
```diff
@@ -136,18 +133,7 @@ func TestCommitLogIndexBootstrap(t *testing.T) {
...
-		SetResultOptions(bsOpts).
-		SetCommitLogOptions(commitLogOpts)
-	fsOpts := setup.storageOpts.CommitLogOptions().FilesystemOptions()
-	bs, err := bcl.NewCommitLogBootstrapperProvider(
-		bclOpts, mustInspectFilesystem(fsOpts), noOpAll)
-	require.NoError(t, err)
-	process := bootstrap.NewProcessProvider(
-		bs, bootstrap.NewProcessOptions(), bsOpts)
-	setup.storageOpts = setup.storageOpts.SetBootstrapProcessProvider(process)
+	setupCommitLogBootstrapperWithFSInspection(t, setup, commitLogOpts)
```
Comment (prateek) [inline src/dbnode/integration/commitlog_bootstrap_index_test.go]: +1 nice

File: src/dbnode/integration/fs_commitlog_mixed_mode_read_write_test.go [hunk 7]
```diff
@@ -244,9 +244,15 @@ func setCommitLogAndFilesystemBootstrapper(t *testing.T, opts testOptions, setup
 	fsBootstrapper, err := fs.NewFileSystemBootstrapperProvider(bfsOpts, commitLogBootstrapper)
 	require.NoError(t, err)
 
+	// Need to make sure we have an active m3dbAdminClient because the previosu one
```
Comment (prateek) [inline src/dbnode/integration/fs_commitlog_mixed_mode_read_write_test.go]: typo in the spelling of previous.

File: src/dbnode/integration/setup.go [hunk 8]
```diff
@@ -574,6 +549,96 @@ func (ts *testSetup) sleepFor10xTickMinimumInterval() {
...
+	id,
+	tchannelNodeAddr string,
+) (client.AdminClient, client.AdminClient, error) {
+	var (
+		clientOpts = defaultClientOptions(topoInit).
+				SetClusterConnectTimeout(opts.ClusterConnectionTimeout()).
+				SetWriteConsistencyLevel(opts.WriteConsistencyLevel()).
+				SetTopologyInitializer(topoInit)
+
+		// We need two distinct clients where one has the origin set to the same ID as the
```
Comment (prateek) [inline src/dbnode/integration/setup.go]: +1 for the explanation. How would you feel about moving it to the struct definition

Comment (richardartoul) [reply src/dbnode/integration/setup.go]: sure

File: src/dbnode/integration/setup.go [hunk 9]
```diff
@@ -574,6 +554,91 @@ func (ts *testSetup) sleepFor10xTickMinimumInterval() {
...
+}
+
+func (ts *testSetup) httpDebugAddr() string {
+	if addr := ts.opts.HTTPDebugAddr(); addr != "" {
+		return addr
+	}
+	return *httpDebugAddr
+}
+
+func (ts *testSetup) maybeResetClients() error {
```
Comment (robskillington) [inline src/dbnode/integration/setup.go]: Super nit: Maybe just call this `resetClients()`?  Kind of implementation detail that it won't if it wasn't set.

**Pull request merged**